yang/frr-isisd.yang

```diff
@@ -188,17 +187,14 @@ module frr-isisd {
           "Applies the conditions of the specified route-map to routes that
            are redistributed into this routing instance.";
       }
-
       leaf metric {
         type uint32 {
           range "0..16777215";
         }
-        default "0";
```

Comment:
Why are you removing this?

Reply:
redistribute for ripd not have it (i created analog), and with `default "0"` i have start isis with:
````
2019/03/29 18:25:56 warnings: ISIS: [EC 100663326] unneeded 'delete' callback for '/frr-isisd:isis/instance/default-information-originate/ipv4/metric'
2019/03/29 18:25:56 warnings: ISIS: [EC 100663326] unneeded 'delete' callback for '/frr-isisd:isis/instance/default-information-originate/ipv6/metric'
2019/03/29 18:25:56 warnings: ISIS: [EC 100663326] unneeded 'delete' callback for '/frr-isisd:isis/instance/redistribute/ipv4/metric'
2019/03/29 18:25:56 warnings: ISIS: [EC 100663326] unneeded 'delete' callback for '/frr-isisd:isis/instance/redistribute/ipv6/metric'
````

Reply:
Hm. Since the `choice` statement was removed and metric has a default value, `metric` will always exist now. In this case, what you need to do is to remove those four `delete` callbacks as they aren't necessary anymore.

Reply:
if i revert `default 0`, i can not write  `redistribute ipv4 static level-1`, or `redistribute ipv4 static level-1 metric 0` in vtysh. Without  `default 0` all ok - and if i create `redistribute ipv4 static level-1 metric 0` or `redistribute ipv4 static level-1`, in config i see `redistribute ipv4 static level-1`. 

Default value of metric =  "0" already exist in  `void redistribute_apply_finish(const struct lyd_node *dnode, int family)`

Reply:
> if i revert default 0, i can not write redistribute ipv4 static level-1, or redistribute ipv4 static level-1 metric 0 in vtysh.

To fix this you'll need to update the `isis_redistribute_cmd` command like this:
```diff
@@ -1101,9 +1101,8 @@ DEFPY(isis_redistribute, isis_redistribute_cmd,
 		nb_cli_enqueue_change(vty, "./route-map",
 				      route_map ? NB_OP_MODIFY : NB_OP_DESTROY,
 				      route_map ? route_map : NULL);
-		nb_cli_enqueue_change(vty, "./metric",
-				      metric ? NB_OP_MODIFY : NB_OP_DESTROY,
-				      metric ? metric_str : NULL);
+		nb_cli_enqueue_change(vty, "./metric", NB_OP_MODIFY,
+				      metric_str ? metric_str : NULL);
 	}
 
 	return nb_cli_apply_changes(
```

Since `metric` is a mandatory leaf now, we can't remove it anymore with `NB_OP_DESTROY`.

> if i create redistribute ipv4 static level-1 metric 0 or redistribute ipv4 static level-1, in config i see redistribute ipv4 static level-1

Good point. You'll need this diff as well to fix the display of the "redistribute" command:
```diff
@@ -1111,13 +1111,13 @@ DEFPY(isis_redistribute, isis_redistribute_cmd,
 }
 
 static void vty_print_redistribute(struct vty *vty, struct lyd_node *dnode,
-                                  const char *family)
+                                  bool show_defaults, const char *family)
 {
        const char *level = yang_dnode_get_string(dnode, "./level");
        const char *protocol = yang_dnode_get_string(dnode, "./protocol");
 
        vty_out(vty, " redistribute %s %s %s", family, protocol, level);
-       if (yang_dnode_exists(dnode, "./metric"))
+       if (show_defaults || !yang_dnode_is_default(dnode, "./metric"))
                vty_out(vty, " metric %s",
                        yang_dnode_get_string(dnode, "./metric"));
        if (yang_dnode_exists(dnode, "./route-map"))
@@ -1129,12 +1129,12 @@ static void vty_print_redistribute(struct vty *vty, struct lyd_node *dnode,
 void cli_show_isis_redistribute_ipv4(struct vty *vty, struct lyd_node *dnode,
                                     bool show_defaults)
 {
-       vty_print_redistribute(vty, dnode, "ipv4");
+       vty_print_redistribute(vty, dnode, show_defaults, "ipv4");
 }
 void cli_show_isis_redistribute_ipv6(struct vty *vty, struct lyd_node *dnode,
                                     bool show_defaults)
 {
-       vty_print_redistribute(vty, dnode, "ipv6");
+       vty_print_redistribute(vty, dnode, show_defaults, "ipv6");
 }
 
 /*
```

For more details about the `show_defaults` parameter, please check this link: https://github.com/opensourcerouting/frr/wiki/Retrofitting-Configuration-Commands#step-6-implement-the-cli_show-callbacks